File: leetcode/3.longest-substring-without-repeating-characters.py
# @leet start
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        last_idx: dict[str, int] = {}

        best = 0
        left = 0

        for right, ch in enumerate(s):
            if ch in last_idx:
                left = max(left, last_idx[ch] + 1)

            last_idx[ch] = right
            best = max(best, right - left + 1)

        return best


# @leet end

File: leetcode/3.longest-substring-without-repeating-characters.ts
// @leet start
function lengthOfLongestSubstring(s: string): number {
  const lastIdx = new Map<string, number>();
  let start = 0;
  let longest = 0;

  for (let i = 0; i < s.length; i++) {
    if (lastIdx.has(s[i])) {
      start = Math.max(start, lastIdx.get(s[i])! + 1);
    }

    longest = Math.max(longest, i - start + 1);
    lastIdx.set(s[i], i);
  }

  return longest;
}
// @leet end
